### PR TITLE
dashboard refactor pt4

### PR DESCRIPTION
**Pull request by mlippert**
### Moved all Metrics components into a `Metrics` subfolder of Dashboard
`ChartCard`, `TurnChart`, `InfluenceChart`, `TimelineChart`, `gantt`

### Extract child components (existing or new) into their own source files
All use the the new common style for components.
`ChartCardInfo`, `MeetingList`, `MeetingListElement`, `WavyBackground`

### styled components
moved styled components into a single `styled.js` file in the same directory as the component(s) that use them.

### Moved relevant redux state connections from the Dashboard to each Metric component
The `index.js` file in the Metrics folder does all of the redux state connecting for all components in that folder, and exports all of the connected components for use by the Dashboard component.

This also means that each metric handles it's data being loaded asynchronously and displays a loading view, until it is available.

This allowed updating the `loadMeetingData` dashboard action creator, to now update the status of each piece of data as it is loaded, instead of waiting for ALL of them to complete.

### TimelineChart / gantt
- Intensive rewrite of the d3 gantt chart class, providing `draw`, `erase` and `redraw` methods (redraw is currently a poor man's implementation, ie it erases and then draws again, but it seems fine for now).
- The revamped gantt class, fixed several issues w/ the TimelineChart (some of which previous refactor may have introduced). Such as clicking on the info button and then closing it added another chart below the existing chart(s). It also allowed code in the TimelineChart to handle a window resize so that the gantt chart could be redrawn to fit in the new size.

### ChartCard
- Replace the 2 specialized properties (`chartDiv` and `chartTable`) by just rendering all children nodes.

### ChartTable
This component is really for accessibility (a11y) so it was moved out of Dashboard and into it's own `A11y` components folder

### colors
More work to use color names defined in libs/utils `Colors`, so it is easier to see and find all the places a color is used, and to know that the color being used is the same as a color used somewhere else.

### fix style issues
A few files were touched to use Color names or just came into my view, and therefore I took the opportunity to fix eslint reported style issues in those files. `App.js`, `Home.js`, `LeaveRoomButton.jsx`, `MakeMeetingCardView.js`, `MakeMeetingCard/index.js`

### Other
- small modifications to eslint config
- added eslint config override for the Dashboard component files (differences like indentation for when they will be extracted into a new repo/package)
- fixed a bulma issue that was being reported when building

### Work that still should be done
- Extract the MeetingInfo that is displayed in the Dashboard to its own component (probably in Metrics). Hook it up to the redux data, and status it needs, so those can be removed from the Dashboard.
- The TimelineChart (and gantt chart) use the task name (namely the user's name) as the link to the task data (utterances) displayed. This causes an issue when multiple participants have the same name. We should use the participant Id as the link, while still displaying the name.
- The TimelineChart size really should be easier to configure w/ props, right now we didn't set it, instead just relying on hard-coded computed values. I'm not quite sure how to do that, ie should the size be just about the gantt chart, or about the entire TimelineChart, ChartCard and all?

## Motivation and context
This is part of the work started in _Dashboard: Refactoring for modularity_ #83 and being continued in _Dashboard: Continue Refactoring for modularity_ #112

## How has this been tested?
Tested locally on my developer machine using `make up` from riff-docker.

## Types of changes

- ✅ Code refactor (non-breaking change)

## Checklist:

- ✅ My code follows the code style of this project. **[REQUIRED]**
- ✅ All new and existing tests passed. **[REQUIRED]**

### CODE REVIEW

File: src/components/Dashboard/Metrics/TimelineChart.jsx [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+ *
+ * React component to visualize the timeline of a Riff meeting.
+ *
+ ********************************************************************************/
+class TimelineChart extends React.PureComponent {
+    static propTypes = {
+
+        /** meeting whose timeline data will be in timelineData */
+        meeting: PropTypes.shape({
+            _id: PropTypes.string.isRequired,
```
Comment (jordanreedie) [inline src/components/Dashboard/Metrics/TimelineChart.jsx]: I'm not super clear on exactly how PropTypes.shape works (and a quick google has not elucidated this for me), does this mean that the meeting object that's being passed *requires* an `_id` property, or that the only property the meeting object can have is `_id`? If the latter, why the change from string to object?

File: src/components/Dashboard/Metrics/TimelineChart.jsx [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+                if (p.id !== participantId) {
+                    return true;
+                }
+                participantMap[p.id] = { name: p.name, color: PeerColors[0] };
+                return false;
+            })
+            .reduce((pmap, p, i) => {
+                pmap[p.id] = {
+                    name: p.name,
+                    color: p.id === participantId ? PeerColors[0] : PeerColors[(i % (PeerColors.length - 1)) + 1],    // eslint-disable-line no-extra-parens
```
Comment (jordanreedie) [inline src/components/Dashboard/Metrics/TimelineChart.jsx]: for the condition here - wasn't `participantId` filtered out already?

File: src/components/Dashboard/Metrics/TimelineChart.jsx [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+                };
+                return pmap;
+            }, participantMap);
+
+        const numOtherPeers = sortedParticipants.length - (participantMap[participantId] ? 1 : 0);
```
Comment (jordanreedie) [inline src/components/Dashboard/Metrics/TimelineChart.jsx]: when would the participantMap for a meeting the user participated in not include the current user? Is `sortedParticipants` derived from utterances and this is for the case where the user doesn't speak at all in a meeting?

File: src/components/Dashboard/Metrics/TimelineChart.jsx (hunk 1 above)
Comment (mlippert) [reply src/components/Dashboard/Metrics/TimelineChart.jsx]: I have to say the prop-types documentation is really lacking. I believe it's "he meeting object that's being passed requires an _id property". I'm pretty sure it doesn't care about any other properties of the meeting object. Now if you reference one of them I'd hope that you'd get a warning that it wasn't defined in propTypes, but I don't know if you would.
The change from just passing the meeting id to passing the meeting object is because what is selected (and in the redux state) is the meeting object. And perhaps there will be other properties of a meeting object it would be desirable to reference (I realize that is a very slippery slope, but I am sort of anticipating that the meeting object will provide its participants at some point)

File: src/components/Dashboard/Metrics/WavyBackground.jsx [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+ * ******************************************************************************/
+
+import React from 'react';
+import PropTypes from 'prop-types';
+
+import { Colors } from 'libs/utils';
+
+import { StyledWave } from './styled';
+
+/* ******************************************************************************
```
[the remark this answers is not in the document]
Comment (mlippert) [reply src/components/Dashboard/Metrics/WavyBackground.jsx]: The file comment and the class comment are 2 different things with different purposes.
It is very true, particularly in a file whose purpose is defining a single class that the description for the file and the class are similar (or even identical). That doesn't mean (to me) that you should delete one of them.

I believe every file should have a file comment with a short summary, possibly more exposition about the file, including the authors, the creation date and the copyright notice.

And every class should also have a class header comment describing its purpose. It can be tough to say a whole lot, and particularly for small classes there just isn't much to say, but the header should still exist.

It provides the easiest source to understand the purpose of the class when all you want is an overview w/o digging into the actual implementation. (That pretty much goes for function/method comments also). Function method block comments (for me) also make it easy to scroll through a source file and see all the functions. When there are no block comments it's easy to miss them. That's why my style looks like it does.

File: src/components/Dashboard/Metrics/TimelineChart.jsx [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+                <div>
+                    <ScaleLoader color={Colors.lightRoyal}/>
+                </div>
+            );
+
+            // TODO:
+            //  - the size of the chart should be specified in a prop somehow, not sure
+            //    if the size should be for the whole chart, chart card and all or just the
+            //    chartDiv. -mjl 2019-05-31
+            return (
```
[the remark this answers is not in the document]
Comment (mlippert) [reply src/components/Dashboard/Metrics/TimelineChart.jsx]: That makes sense, but I like the explicitness of 2 return paths in this situation, because I think of the loading path as "exceptional" and the path that displays the chart as the normal case, and I wanted that to be obvious.

I have been thinking about how to name a method to return the loading jsx, in order to make that "exception" case smaller in the render, I just hadn't gotten there yet. And using a single var like `content` may be the way to go also.

There is definitely more "refactor" to be done to get the jsx looking as clean as you've got in chat (particularly in the dashboard itself).

File: src/components/Dashboard/Metrics/WavyBackground.jsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (jordanreedie) [reply src/components/Dashboard/Metrics/WavyBackground.jsx]: Ah, ok. I'm personally not a super big fan of having a class comment when we have the file comment if we're sticking to one component per file. IMO, unless there is additional information that needs to be conveyed, it just adds an extra thing to maintain and potentially get out of sync & cause confusion. That being said, I understand why you want them and I don't have particularly strong feelings about it so I'll concede this one

File: src/components/Dashboard/Metrics/TimelineChart.jsx [hunk 6]
```diff
@@ -0,0 +1,270 @@
...
+import PropTypes from 'prop-types';
+import { ScaleLoader } from 'react-spinners';
+
+import { Colors, PeerColors, logger } from 'libs/utils';
+
+import { Gantt } from './gantt';
+import { ChartCard } from './ChartCard';
+import { WavyBackground } from './WavyBackground';
+
+const chartInfo = 'This chart shows a timeline of when people spoke during your meeting.';
```
Comment (jordanreedie) [inline src/components/Dashboard/Metrics/TimelineChart.jsx]: oh also why is `chartInfo` declared up here but `chartTitle` declared in the `render` function? It looks like they're both used in the same places

File: src/components/Dashboard/Metrics/TimelineChart.jsx (hunk 2 above)
Comment (mlippert) [reply src/components/Dashboard/Metrics/TimelineChart.jsx]: I didn't revisit this code in this latest refactor, but yes you're right (that check is an unnecessary hold over and should be removed.

File: src/components/Dashboard/Metrics/TimelineChart.jsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (jordanreedie) [reply src/components/Dashboard/Metrics/TimelineChart.jsx]: Oh ok, I wasn't sure if this was the last refactor pass or not, I don't think you need to address it immediately.

File: src/components/Dashboard/Metrics/TimelineChart.jsx (hunk 1 above)
Comment (jordanreedie) [reply src/components/Dashboard/Metrics/TimelineChart.jsx]: Ok, that's fair enough!

File: src/components/Dashboard/Metrics/TimelineChart.jsx (hunk 3 above)
Comment (mlippert) [reply src/components/Dashboard/Metrics/TimelineChart.jsx]: Yes, I'm pretty sure that is what's going on. Passing better data is still on the to do list.

File: src/components/Dashboard/Metrics/TimelineChart.jsx (hunk 6 above)
Comment (mlippert) [reply src/components/Dashboard/Metrics/TimelineChart.jsx]: That is a very good question. And they should be consistent.

I think it is more efficient to declare private constants at the module level, rather than in a function. (I'm not sure about that).

I probably was trying to see how I felt about the module level constants, and ended up doing things inconsistently. Which way do you think they should go? I'm currently leaning towards moving the `chartCardTitle` to the module level next to `chartInfo`.

File: src/components/Dashboard/Metrics/TimelineChart.jsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (mlippert) [reply src/components/Dashboard/Metrics/TimelineChart.jsx]: I was just looking at it again, And the same way it was bugging you, it's been bugging me. I prefer the 2 return paths, but I'd like a method that takes the "content" and returns in wrapped in that common stuff. I just haven't figure out how to name that method yet.

File: src/components/Dashboard/Metrics/WavyBackground.jsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (mlippert) [reply src/components/Dashboard/Metrics/WavyBackground.jsx]: Thanks :smile:

File: src/components/Dashboard/Metrics/TimelineChart.jsx (hunk 6 above)
Comment (jordanreedie) [reply src/components/Dashboard/Metrics/TimelineChart.jsx]: Hmmm I could probably be swayed either way, I think. When you say 'more efficient', do you mean for performance or human-readability? I can't really imagine it having much of an effect on actual performance. 

The reason I noticed that these were in different places is because when I was looking over the render code, I realized I had no idea where `chartInfo` was coming from. So, I think for readability's sake, it might be better to have it in the render function? I think that's consistent with most of our other code also.

Comment (mlippert) [reply src/components/Dashboard/Metrics/TimelineChart.jsx]: I did mean performance, and yes in this case it is totally something that would be negligible, I just hate the idea that something that only needs to be created once, might be created every time a function is run.

The TurnChart actually defines both constants at the module level.

File: src/components/Dashboard/Metrics/TimelineChart.jsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (jordanreedie) [reply src/components/Dashboard/Metrics/TimelineChart.jsx]: Haha yeah just looking at it, my immediate expectation is the two return paths must be totally different, and I think it's the subversion of that expectation that's annoying. I can't think of a good name for it either, though

File: src/components/Dashboard/Metrics/TimelineChart.jsx (hunk 6 above)
Comment (mlippert) [reply src/components/Dashboard/Metrics/TimelineChart.jsx]: OK after some discussion w/ @jordanreedie in slack, the info and title properties are now in a `chartConfig` module level object in both TurnChart and TimelineChart.

**Comment by mlippert** [comment]
OK, I've committed updates so that the TurnChart and TimelineChart now define a module level `chartConfig` with cardTitle and info properties.
And I've removed that redundant check of the participant ID when assigning other's colors in the timeline.
And lastly I added a `_decorateContent` method in both the TurnChart and TimelineChart so that the decoration elements aren't duplicated in the `render` implementation.

File: src/components/Dashboard/Metrics/gantt.js [hunk 7]
```diff
@@ -28,152 +39,166 @@ const TimeDomainMode = {
...
 
     get width() {
-        return this.privateProp.width;
+        return this._privateProp.width;
     }
     set width(w) {
-        this.privateProp.width = Number(w);
+        let newWidth = w;
+        if (!newWidth) {
+            newWidth = document.body.clientWidth / 2;
```
Comment (jaedoucette) [inline src/components/Dashboard/Metrics/gantt.js]: Same expr.

File: src/components/Dashboard/Metrics/gantt.js [hunk 8]
```diff
@@ -387,7 +416,7 @@ function wrapYLabels(yLabels, width) {
         // otherwise put it on the next line and make that the current line
         let numLines = 1;
         let word;
-        while (word = words.pop()) {
+        while (word = words.pop()) {    // eslint-disable-line no-cond-assign
```
Comment (jaedoucette) [inline src/components/Dashboard/Metrics/gantt.js]: Consider changing this to a for loop to avoid the eslint and to make more explicit?

**Comment by jaedoucette** [comment]
> These are comments Brec and I's joint review.

@brecriffs

File: src/components/Dashboard/ChartCard.jsx [hunk 9]
```diff
@@ -14,15 +14,10 @@
  *
  * ******************************************************************************/
 
-// Set eslint indentation to 4 instead of 2 in preparation to moving this to a new
-// repository, where that will be the rule.
 // I haven't figured out for sure if we(Riff) want to use the underscore prefix to
```
[the remark this answers is not in the document]
Comment (mlippert) [reply src/components/Dashboard/ChartCard.jsx]: I'm not ready to do that, because I'm still on the fence about the style. If we decide that we want to use the '_' prefix to denote class properties that are considered private, then I'd move that rule to the eslint config for the src tree (and probably also to the config for the server tree). Having the override in particular files makes it easy to find where it being used, and fix it if we decide against it.

File: src/components/Dashboard/DashboardView.jsx [hunk 10]
```diff
@@ -74,9 +73,53 @@ class MeetingView extends React.Component {
         handleMeetingClick: PropTypes.func.isRequired,
     }
 
+    /**
```
[the remark this answers is not in the document]
Comment (mlippert) [reply src/components/Dashboard/DashboardView.jsx]: It's not general purpose enough in its current form to be in utils, it really does something very specific for this class.

File: src/components/Dashboard/index.js [hunk 11]
```diff
@@ -61,15 +60,18 @@ const mapDispatchToProps = dispatch => ({
...
 const formatMeetingDuration = (meeting) => {
     if (meeting === null) {
         return '';
     }
     // support showing data on 'in progress' meeting by giving it a fake end time of now
     const startDate = new Date(meeting.startTime);
     const endDate = meeting.endTime ? new Date(meeting.endTime) : new Date();
-    const diff = moment(endDate).diff(moment(startDate), 'minutes');
-    return `${diff} minutes`;
+    const durationMins = Math.trunc((endDate.getTime() - startDate.getTime()) / MILLISEC_PER_MINUTE);
```
[the remark this answers is not in the document]
Comment (mlippert) [reply src/components/Dashboard/index.js]: Hmm, it doesn't strike me as particularly less readable, and it some ways I find the new code easier to understand what is being done, so I think we'll just disagree on this one.

File: src/components/Dashboard/index.js [hunk 12]
```diff
@@ -72,11 +71,10 @@ const formatMeetingDuration = (meeting) => {
     if (meeting === null) {
         return '';
     }
-    // support showing data on in progress meeting by giving them a fake end time of now
-    if (!meeting.endTime) {
-        meeting = { ...meeting, endTime: new Date() };
-    }
-    const diff = moment(new Date(meeting.endTime)).diff(moment(new Date(meeting.startTime)), 'minutes');
+    // support showing data on 'in progress' meeting by giving it a fake end time of now
```
[the remark this answers is not in the document]
Comment (mlippert) [reply src/components/Dashboard/index.js]: good point. I'll update the comment.

File: src/components/Dashboard/.editorconfig [hunk 13]
```diff
@@ -0,0 +1,19 @@
+# http://editorconfig.org
```
[the remark this answers is not in the document]
Comment (mlippert) [reply src/components/Dashboard/.editorconfig]: It's completely different from an eslint configuration file, did you read the website?
It's a hint file for editors if they support it to set things like use tabs or spaces for various files in the hierarchy. It doesn't have anything to do with how you actually end up structuring your code.

There is an .editorconfig file already at the top level, I added this one here for the same reason as I added the override eslint config, I want to use different indentation for the files in Dashboard, and I was tired of my editor setting the indentation differently than what I wanted to use.

File: src/components/Dashboard/ChartCard.jsx (hunk 9 above)
[the remark this answers is not in the document]
Comment (jaedoucette) [reply src/components/Dashboard/ChartCard.jsx]: That makes sense. I did some more reading, and it looks like as of the end of March, the JS standards support `#` as an interpreter-enforced privacy prefix. Apparently Chrome's interpreter already ships with this feature, and presumably others will follow. 

In light of that, maybe we should migrate to the `#`-prefix as well?

See: https://github.com/tc39/proposal-class-fields/blob/master/PRIVATE_SYNTAX_FAQ.md

and 

https://stackoverflow.com/questions/4484424/underscore-prefix-for-property-and-method-names-in-javascript

[the remark this answers is not in the document]
Comment (jaedoucette) [reply src/components/Dashboard/ChartCard.jsx]: That makes sense. I did some more reading, and it looks like as of the end of March, the JS standards support `#` as an interpreter-enforced privacy prefix. Apparently Chrome's interpreter already ships with this feature, and presumably others will follow. 

In light of that, maybe we should migrate to the `#`-prefix as well?

See: https://github.com/tc39/proposal-class-fields/blob/master/PRIVATE_SYNTAX_FAQ.md

and the accepted answer to:

https://stackoverflow.com/questions/4484424/underscore-prefix-for-property-and-method-names-in-javascript

[the remark this answers is not in the document]
Comment (jaedoucette) [reply src/components/Dashboard/ChartCard.jsx]: That makes sense. I did some more reading, and it looks like as of the end of March, the JS standards support `#` as an interpreter-enforced privacy prefix. Apparently Chrome's interpreter already ships with this feature, and presumably others will follow. 

In light of that, maybe we should migrate to the `#`-prefix as well?

See: https://github.com/tc39/proposal-class-fields/blob/master/PRIVATE_SYNTAX_FAQ.md

and the accepted answer to:

https://stackoverflow.com/questions/4484424/underscore-prefix-for-property-and-method-names-in-javascript

File: src/.eslintrc.yml [hunk 14]
```diff
@@ -179,6 +179,8 @@ rules:
     - error
     -
       functions: false
+      classes: false
```
[the remark this answers is not in the document]
Comment (mlippert) [reply src/.eslintrc.yml]: In general you shouldn't be executing anything in a module, therefore a forward reference within a function to something defined later in the module should be safe, and is desirable occasionally, more often than the message not to do that is helpful in my opinion.

File: src/components/Dashboard/DashboardView.jsx (hunk 10 above)
[the remark this answers is not in the document]
Comment (jaedoucette) [reply src/components/Dashboard/DashboardView.jsx]: It's not clear from the code snippet attached to this comment, but I think we meant the data/time functionality that got pulled out of the moment package. Looking again, it seems pretty general, and also pretty unrelated to the class (apart from the fact that only this class wants that functionality right now?). 

Ultimately fine if it is specific to this class, but wanted to make sure we were talking about the same thing.

File: src/components/Dashboard/ChartCard.jsx (hunk 9 above)
[the remark this answers is not in the document]
Comment (mlippert) [reply src/components/Dashboard/ChartCard.jsx]: When ecmascript officially supports a method of specifying private fields we should definitely use that. We can't use the `#` prefix now, as it isn't a valid character for a variable name.

File: src/components/Dashboard/DashboardView.jsx (hunk 10 above)
[the remark this answers is not in the document]
Comment (mlippert) [reply src/components/Dashboard/DashboardView.jsx]: Well this comment is on code that isn't in the final commit in this file, but is in the `MeetingListElement.jsx` file where there is a static method `_formatMeetingTime`.
I don't consider formatting a date/time to a very specific string format, to be general purpose, and yes this is the only place that format of string is used. Some of the other functions are almost general purpose enough to warrant being extracted to the utils lib, but I don't think they're quite ready, I'd want to see them needed somewhere else first.

File: src/components/Dashboard/MeetingList.jsx [hunk 15]
```diff
@@ -0,0 +1,108 @@
...
+                {this._getMeetingTiles()}
+            </MeetingTabs>
+        );
+    }
+
+    /* **************************************************************************
+     * _getMeetingTiles (private)                                          */ /**
+     *
+     * Returns an array of MeetingListElements for all of the meetings that
+     * belong in this MeetingList.
```
[the remark this answers is not in the document]
Comment (mlippert) [reply src/components/Dashboard/MeetingList.jsx]: Yes all meetings supplied in the prop belong in the list. How would you word the method description?

File: src/components/Dashboard/MeetingList.jsx [hunk 16]
```diff
@@ -0,0 +1,108 @@
...
+     *
+     * @returns {Array<MeetingListElement>}
+     */
+    _getMeetingTiles() {
+        return this.props.meetings
+            .map((meeting) => {
```
[the remark this answers is not in the document]
Comment (mlippert) [reply src/components/Dashboard/MeetingList.jsx]: really? `selected` isn't clearly a boolean indicating whether the current meeting is the selectedMeeting?

File: src/components/Dashboard/MeetingListElement.jsx [hunk 17]
```diff
@@ -0,0 +1,166 @@
...
+     * ha - hour am/pm; MMM - 3 letter month abbrev.; Do - Day of Month 1st 2nd ... 30th 31st
+     * e.g. returns a string like: '10am Feb 21st'.
+     * if we need to get much fancier, such as localizing the month abbreviations, that would
+     * be worth including the moment package.
+     *
+     * @param {Date} dt
+     *
+     * @returns {string} a formatted representation of the given date
+     */
+    static _formatMeetingTime(dt) {
```
[the remark this answers is not in the document]
Comment (mlippert) [reply src/components/Dashboard/MeetingListElement.jsx]: see earlier discussion

File: src/components/Dashboard/MeetingListElement.jsx [hunk 18]
```diff
@@ -0,0 +1,166 @@
...
+        const dayOfMonth = dt.getDate();
+        return `${getHoursAmPm(dt)} ${monthAbbrs[dt.getMonth()]} ${dayOfMonth}${ordinalIndicator(dayOfMonth)}`;
+    }
+}
+
+/** Array of english month abbreviations */
+const monthAbbrs = [ 'Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec' ];
```
[the remark this answers is not in the document]
Comment (mlippert) [reply src/components/Dashboard/MeetingListElement.jsx]: All of this is very specific to creating a particular string for a particular purpose. Certainly the moment formatter is more succint, and potentially clearer, but I really wanted to try to remove the couple of small uses we had for that library which is very large.

File: src/components/Dashboard/DashboardView.jsx [hunk 19]
```diff
@@ -133,7 +152,8 @@ class DashboardView extends React.Component {
                 </div>
             );
         }
-        else if (fetchMeetingsStatus === 'error') {
+
+        if (fetchMeetingsStatus === 'error') {
```
[the remark this answers is not in the document]
Comment (mlippert) [reply src/components/Dashboard/DashboardView.jsx]: I think it's clearer that previous condition was handled there is no path that continues, which isn't true if you still have the else.
I like the style of handle simple cases/error conditions and  return up front with the more normal "interesting" functionality following that.
The mattermost team had turned on the lint to flag else after an if that returned, and I agreed with that as a general practice.
I could see converting this particular set of tests to it's own table lookup possibly in a future refactor, although with only 3 cases I'm not sure it would be worth it.

File: src/components/Dashboard/DashboardView.jsx [hunk 20]
```diff
@@ -145,12 +165,86 @@ class DashboardView extends React.Component {
...
-        // is ugly and not what the user saw when joining the room so remove that course id suffix
+        // The real room name of LTI user meetings in the meeting object has the
+        // course id as a suffix which is ugly and not what the user saw when joining
+        // the room so remove that course id suffix
         const selectedRoomName = !selectedMeeting ? null
                                                   : isLtiUser ? selectedMeeting.room.split('_').slice(0, -1).join('_')  // eslint-disable-line indent
                                                               : selectedMeeting.room;                                   // eslint-disable-line indent
 
+        let meetingInfo;
+        if (!selectedMeeting || statsStatus !== 'loaded') {
```
[the remark this answers is not in the document]
Comment (mlippert) [reply src/components/Dashboard/DashboardView.jsx]: It shouldn't be possible to have loaded stats if a meeting isn't selected, because I'm not sure what that would even mean. But the way that some of these properties are updated in the redux store I felt it's better to be safe if that occurs rather than access an undefined variable. How stuff is represented and updated in the store is something for a future refactor.

Note that this is only for the meeting info, to be displayed or not displayed. And, this is actually one of those places that I think needs additional refactoring, which would probably clear some of this up. The MeetingInfo really needs to be extracted into a separate component connected to the appropriate redux properties. This is definitely making the code more complex than it should be. But that isn't happening right now.

File: src/components/Dashboard/MeetingListElement.jsx [hunk 21]
```diff
@@ -54,11 +51,10 @@ class MeetingListElement extends React.Component {
      * @see {@link https://reactjs.org/docs/react-component.html#render|React.Component.render}
      */
     render() {
-        const { uid, meeting, selected, handleMeetingClick } = this.props;
```
[the remark this answers is not in the document]
Comment (mlippert) [reply src/components/Dashboard/MeetingListElement.jsx]: I felt it was better to see exactly where props where being used, instead of having a local variable created for them. This de-structuring was originally added as an interim step as I refactored from the previous function form of the component.

File: src/components/Dashboard/index.js [hunk 22]
```diff
@@ -26,39 +26,59 @@ import { logger } from 'libs/utils';
...
-    },
-    handleMeetingClick: (event, uid, meeting) => {
-        event.preventDefault();
-        logger.debug('selected meeting', meeting._id);
-        /// TODO: Why do we want 2 distinct actions to select the meeting and to load the meeting data? -mjl 2018-09-05
-        dispatch(selectMeeting(meeting));
-        dispatch(loadMeetingData(uid, meeting._id));
-    },
-});
+const dashboardMapProps = {
```
[the remark this answers is not in the document]
Comment (mlippert) [reply src/components/Dashboard/index.js]: sorry, I'm not sure what this comment is referring to?

I ended up getting rid of the connectedComponent function because I didn't feel the efficiency of the using the function was worth the mental effort to follow the indirection to understand what was being done. Particularly if that function got moved to the utility lib. Even in the case where there were multiple components being connected (`Metrics/index.js`)

File: src/components/Dashboard/MeetingListElement.jsx [hunk 23]
```diff
@@ -52,37 +52,43 @@ class MeetingListElement extends React.Component {
      */
     render() {
         const m = MeetingListElement._formatMeetingTime(new Date(this.props.meeting.startTime));
+        const itemClasses = [ 'timeline-item' ];
+        let markerColor = '#576066'; // not selected
```
[the remark this answers is not in the document]
Comment (mlippert) [reply src/components/Dashboard/MeetingListElement.jsx]: I'm pretty sure I did that.

File: src/components/MakeMeetingCard/MakeMeetingCardView.jsx [hunk 24]
```diff
@@ -11,51 +27,85 @@ border-radius: 5px;
 .card-content p {
   color: #fff;
 }
-`
+`;
 
 // linear-gradient(30deg, rgba(84,48,96,1) 20%, rgba(138,106,148,1) 50%, rgba(229,178,245,1) 100%);
-//linear-gradient(66deg, rgba(70,19,85,1) 0%, rgba(138,106,148,1) 69%, rgba(201,155,215,0.9) 100%);
+// linear-gradient(66deg, rgba(70,19,85,1) 0%, rgba(138,106,148,1) 69%, rgba(201,155,215,0.9) 100%);
```
[the remark this answers is not in the document]
Comment (mlippert) [reply src/components/MakeMeetingCard/MakeMeetingCardView.jsx]: It was there before I started the refactor, and I'm not sure either, so I was/am not ready to delete it.

File: src/components/MakeMeetingCard/index.js [hunk 25]
```diff
@@ -1,44 +1,51 @@
...
 
-import { changeRoomNameState } from "redux/actions/makeMeeting";
-import { changeRoomName } from "redux/actions/chat";
+import { changeRoomNameState } from 'redux/actions/makeMeeting';
+import { changeRoomName } from 'redux/actions/chat';
 
-import MakeMeetingCardView from './MakeMeetingCardView';
+import { MakeMeetingCardView } from './MakeMeetingCardView';
 
 const mapStateToProps = state => ({
```
[the remark this answers is not in the document]
Comment (mlippert) [reply src/components/MakeMeetingCard/index.js]: This wasn't a refactor. This was just a fly-by fixing of lint errors, probably because I touched the file for some minor reason like using named colors.

The refactoring was all focused solely on Dashboard files.

File: src/components/Dashboard/Metrics/styled.js [hunk 26]
```diff
@@ -0,0 +1,82 @@
...
+import styled from 'styled-components';
+
+import { Colors } from 'libs/utils';
+
+const CardTitle = styled.div.attrs({
+    className: 'title is-5 has-text-left'
+})`
+margin-left: 1rem;
+margin-right: 1rem;
+color: rgb(138,106,148);
```
[the remark this answers is not in the document]
Comment (mlippert) [reply src/components/Dashboard/Metrics/styled.js]: yep, looks like I missed some.

File: src/components/Dashboard/Metrics/TimelineChart.jsx [hunk 27]
```diff
@@ -179,8 +183,29 @@ class TimelineChart extends React.Component {
...
-    _drawGantt() {
+    /* **************************************************************************
+     * _onResize                                                           */ /**
+     *
+     * Handle a resize event of the window, which we want to resize the gantt
+     * chart.
+     */
+    _onResize(e) {
+        logger.debug('TimelineChart._onResize:', e);
+        this.setState({ chartWidth: document.body.clientWidth / 2 });
```
[the remark this answers is not in the document]
Comment (mlippert) [reply src/components/Dashboard/Metrics/TimelineChart.jsx]: Just curious, if it wasn't an "expression" but just document.body.clientWidth would you have suggested moving it to a function?
I think you're right that setting the initial value for the chart width, and updating it should be done in the same place, ie it is worth extracting it to a function.

File: src/components/Dashboard/Metrics/TimelineChart.jsx [hunk 28]
```diff
@@ -109,24 +130,7 @@ class TimelineChart extends React.Component {
      */
     componentDidUpdate() {
         logger.debug('TimelineChart: component updating...', this.props);
-    }
-
-    /* **************************************************************************
-     * shouldComponentUpdate                                               */ /**
```
[the remark this answers is not in the document]
Comment (mlippert) [reply src/components/Dashboard/Metrics/TimelineChart.jsx]: No.
Who had those concerns?
The issues from my 1st pass refactor, and when I added the shouldComponentUpdate returning false had to do with how one goes about utilizing d3 within a react component.

Learning more about [shouldComponentUpdate](https://reactjs.org/docs/react-component.html#shouldcomponentupdate), returning false was working for now, but wasn't really the right way to do this anyway. Giving the gantt chart d3 class the ability to redraw and erase and then utilizing that functionality is the correct solution, and is what I finally got around to here (well it's still a poor man's redraw, but it does the job). See the devnote on the file comment for more info.

File: src/components/Dashboard/Metrics/gantt.js [hunk 29]
```diff
@@ -28,152 +39,166 @@ const TimeDomainMode = {
...
-            bottom: 20,
-            left: 50,
+    constructor() {
+        /** backing store for getter/setter properties, and other private properties
+         *  sets initial values for those properties.
+         */
+        this._privateProp = {
+            parentSelector: null,
+            svg: null,
+            width: document.body.clientWidth / 2,
```
[the remark this answers is not in the document]
Comment (mlippert) [reply src/components/Dashboard/Metrics/gantt.js]: same expression, but different context, so only tangentially related.

File: src/components/Dashboard/Metrics/gantt.js (hunk 7 above)
Comment (mlippert) [reply src/components/Dashboard/Metrics/gantt.js]: however, yes this is a possibly a case of resetting an initial value.

Actually though I think this is somewhat different, and the 2 expressions while the same are not really serving the same purpose. This case is more error handling, w/ a log message to explicitly call that out. The fact that I chose to use the same expression as is currently used to initialize the backing value is happenstance  (well not really, but I could see changing what the initial value is set to and not changing this).
So I'm leaving this as is.

File: src/components/Dashboard/Metrics/gantt.js (hunk 8 above)
Comment (mlippert) [reply src/components/Dashboard/Metrics/gantt.js]: I thought about it, and I'll probably think about it again, but I decided to leave this for now as matching the original source where this function was copied from.

**Comment by mlippert** [comment]
OK, I've made changes based on @jaedoucette and @brecriffs 's review comments  and pushed the commit with them.
Check them out and if things look good now, approve the PR please.

**Comment by jaedoucette** [comment]
It probably wouldn't have caught my attention as something we should look
at more closely without the compound expression, but Brec and I did discuss
this as being primarily a maintainability issue because of the two
locations.

File: src/components/Dashboard/Metrics/TimelineChart.jsx (hunk 28 above)
[the remark this answers is not in the document]
Comment (jaedoucette) [reply src/components/Dashboard/Metrics/TimelineChart.jsx]: The comments preceding the original suggested that this was a performance optimization that had been removed. Brec had a vague recollection of a recent bug related to a component that redrew much too often. We figured we should flag this for a second look.

File: src/components/Dashboard/MeetingList.jsx (hunk 15 above)
[the remark this answers is not in the document]
Comment (mlippert) [reply src/components/Dashboard/MeetingList.jsx]: That's very specific. As I think about it more I thing _belong_ may be the right term after all as that is the purpose of the method, and what is expected where it is being used. The fact that **everything** currently belongs is an implementation detail that I think it is reasonable to read the code if you want to know that. Well maybe what belongs should also be in the description.

OK I've reworked the description and pushed it.

**Comment by mlippert** [comment]
**Release** [1.2.0-dev.9](https://github.com/rifflearning/riff-rtc/releases/tag/1.2.0-dev.9)
